### ISSUE TITLE
ElasticSearch signing request issues

### ISSUE THREAD

**Issue by 0xTim** [soto-project/soto#466]
When attempting to make a request to AWS ElasticSearch I'm getting a 403 Forbidden error, with the response

```json

{
    "message": "The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details."
}
```

This is in a Lambda function. My `ElasticSearchClient` looks like:

```swift
import AsyncHTTPClient
import NIO
import Foundation
import NIOHTTP1
import SotoElasticsearchService

// MARK: Client
public final class ElasticsearchClient {

    private let client: HTTPClient
    private let awsClient: AWSClient
    private let eventLoop: EventLoop
    private let logger: Logger
    private let scheme: String
    private let host: String
    private let port: Int
    private let username: String?
    private let password: String?

    public init(eventLoop: EventLoop, logger: Logger, client: HTTPClient, awsClient: AWSClient, scheme: String = "http", host: String, port: Int = 9200, username: String? = nil, password: String? = nil) {
        self.eventLoop = eventLoop
        self.logger = logger
        self.client = client
        self.awsClient = awsClient
        self.scheme = scheme
        self.host = host
        self.port = port
        self.username = username
        self.password = password
    }

    private func sendRequest<D: Decodable>(_ request: HTTPClient.Request, requestBody: Data?) throws -> EventLoopFuture<D> {
        return try sendRequest(request, requestBody: requestBody).flatMapThrowing { clientResponse in
            self.logger.debug("Response: \(clientResponse)")
            if let responseBody = clientResponse.body {
                self.logger.debug("Response body: \(String(decoding: responseBody.readableBytesView, as: UTF8.self))")
            }
            switch clientResponse.status.code {
            case 200...299:
                guard var body = clientResponse.body else {
                    throw HandlerError(message: "No body from from ElasticSearch response")
                }
                guard let response = try body.readJSONDecodable(D.self, length: body.readableBytes) else {
                    throw HandlerError(message: "Failed to convert \(D.self)")
                }
                return response
            default: throw HandlerError(message: "Got response status \(clientResponse.status) from ElasticSearch")
            }
        }
    }

    private func sendRequest(_ request: HTTPClient.Request, requestBody: Data?) throws -> EventLoopFuture<HTTPClient.Response> {
        var request = request
        let body: AWSPayload
        if let requestBody = requestBody {
            body = .data(requestBody)
            request.body = HTTPClient.Body.data(requestBody)
        } else {
            body = .empty
        }
        let es = ElasticsearchService(client: awsClient)
        return es.signHeaders(url: request.url, httpMethod: request.method, headers: request.headers, body: body, logger: logger).flatMap { headers in
            request.headers = headers
            self.logger.debug("Request: \(request)")
            return self.client.execute(request: request, logger: self.logger)
        }
    }
}

//// MARK: - Helper
extension ElasticsearchClient {

    private func baseURL(path: String, queryItems: [URLQueryItem] = []) -> String {
        var urlComponents = URLComponents()
        urlComponents.scheme = scheme
        urlComponents.host = host
        urlComponents.port = port
        urlComponents.path = path
        urlComponents.queryItems = queryItems
        guard let url = urlComponents.url else {
            fatalError("malformed url: \(urlComponents)")
        }
        return url.absoluteString
    }
}

// MARK: Requests
extension ElasticsearchClient {

    public func createDocument<Document: Encodable>(_ document: Document, in indexName: String) -> EventLoopFuture<ESCreateDocumentResponse> {
        let url = baseURL(path: "/\(indexName)/_doc")
        do {
            var request = try HTTPClient.Request(url: url, method: .POST)
            let body = try JSONEncoder().encode(document)
            request.headers.add(name: "content-type", value: "application/json")
            return try sendRequest(request, requestBody: body)
        } catch {
            return self.eventLoop.makeFailedFuture(error)
        }
    }

    public func updateDocument<Document: Encodable>(_ document: Document, id: String, in indexName: String) -> EventLoopFuture<ESUpdateDocumentResponse> {
        let url = baseURL(path: "/\(indexName)/_doc/\(id)")
        do {
            var request = try HTTPClient.Request(url: url, method: .PUT)
            let body = try JSONEncoder().encode(document)
            request.headers.add(name: "content-type", value: "application/json")
            return try sendRequest(request, requestBody: body)
        } catch {
            return self.eventLoop.makeFailedFuture(error)
        }
    }

    public func deleteDocument(id: String, from indexName: String) -> EventLoopFuture<ESDeleteDocumentResponse> {
        let url = baseURL(path: "/\(indexName)/_doc/\(id)")
        do {
            let request = try HTTPClient.Request(url: url, method: .DELETE)
            return try sendRequest(request, requestBody: nil)
        } catch {
            return self.eventLoop.makeFailedFuture(error)
        }
    }
}
```

The important part is in `sendRequest((_:body:)`. 

I'm invoking it with:

```swift
elasticSearchClient.createDocument(item, in: "items")
```

Where `item` is a type that conforms to `Codable`.

In CloudWatch I can see the outgoing request:

```
Request: Request(method: NIOHTTP1.HTTPMethod.POST, url: https://<ES-ENDPOINT>.us-east-1.es.amazonaws.com:443/items/_doc?, scheme: "https", host: "<ES_ENDPOINT>.us-east-1.es.amazonaws.com", socketPath: "", uri: "/items/_doc", headers: [("content-type", "application/json"), ("X-Amz-Date", "20210201T204236Z"), ("host", "<ES_ENDPOINT>.us-east-1.es.amazonaws.com:443"), ("x-amz-content-sha256", "58a7bcf28bd69a4a512e4fa5a8a6d480e96c156e5460ac9cd7f94e94719124ec"), ("x-amz-security-token", "<LARGE_TOKEN_PAYLOAD>=="), ("Authorization", "AWS4-HMAC-SHA256 Credential=<SOME_ID>/20210201/us-east-1/es/aws4_request, SignedHeaders=content-type;host;x-amz-content-sha256;x-amz-date;x-amz-security-token, Signature=88ce83b2a9dfb3ea866f1<MORESIGNATURE>")], body: Optional(AsyncHTTPClient.HTTPClient.Body(length: Optional(435), stream: (Function))), redirectState: nil, kind: AsyncHTTPClient.HTTPClient.Request.Kind.host)
```

Using versions:

SotoCore and Soto 5.1.0

**Comment by adam-fowler** [comment]
Hey @0xTim 
I resurrected my ElasticSearch code and it is still working. 

So looking through your code the only thing that comes to light is you have not explicitly set the region in your `ElasticsearchService` instance. The region will be automatically taken from the Lambda region in this case. Could you change that line to 
```swift
es = ElasticsearchService(client: awsClient, region: .useast1)
```
given your elastic search instance is in us-east-1

I guess the other thing would be, although I can't see how it would cause an issue, is when you setup the request body for signing you create two different copies of payload, one for the request and one for signing. Both of these internally store `ByteBuffers`. You could create a `ByteBuffer` and set both of them with this `ByteBuffer`.

For reference here is my working code
```swift
func elasticSearchExecute(url: String, method: HTTPMethod, headers: HTTPHeaders, body: AWSPayload = .empty) -> EventLoopFuture<HTTPClient.Response> {
    let es = ElasticsearchService(client: awsClient, region: .useast1)
    return es.signHeaders(
        url: URL(string: url)!,
        httpMethod: method,
        headers: headers,
        body: body
    ).flatMap { headers in
        let request = try! HTTPClient.Request(
            url: url,
            method: method,
            headers: headers,
            body: body.asByteBuffer().map { .byteBuffer($0) }
        )
        return httpClient.execute(request: request)
    }
}
```

**Comment by 0xTim** [comment]
Ok, I've found the issue. The problem is with the port and this line

```swift
urlComponents.port = port
```

This adds the port to the URL, which broke the signing. Removing that line, and letting the client pick the port with the scheme (since AWS just uses 443 and HTTPS) allowed the requests to go through

**Comment by adam-fowler** [comment]
This appears to fix it https://github.com/soto-project/soto-core/pull/421
Need to be sure it doesnt break anything else

**Comment by adam-fowler** [comment]
Fix in 5.1.1 of soto-core
